firewall/inspection/portscan/detect.go

```diff
@@ -0,0 +1,357 @@
+		port := tcpUDPport{protocol: conn.IPProtocol, port: conn.LocalPort}
+		for _, e := range entry.blockedPorts {
+			if e == port {
+				triedPort = true
+				break
+			}
+		}
+
+		if !triedPort {
+			entry.blockedPorts = append(entry.blockedPorts, tcpUDPport{protocol: conn.IPProtocol, port: conn.LocalPort})
+			entry.score = intMin(entry.score+addScore, maxScore)
+
+			if entry.previousOffender || entry.score >= scoreBlock {
+				entry.blocked = true
+				entry.previousOffender = true
+
+				//TODO: actually I just want to know if THIS threat exists - I don't need prefixing. Maybe we can do it simpler ... (less CPU-intensive)
+				if t, _ := status.GetThreats(threadPrefix + ipString); len(t) == 0 {
+					log.Tracer(ctx).Debugf("new Threat")
+					status.AddOrUpdateThreat(&status.Threat{
+						ID:              threadPrefix + ipString,
+						Name:            "Portscan by " + ipString,
+						Description:     "Someone tries to connect to a lot of closed Ports (non-running Services). Probably he wants to find out the services running on the maschine to determine which services to attack", //fixme: to long
+						MitigationLevel: status.SecurityLevelHigh,
+						Started:         time.Now().Unix(),
+					})
+				}
+			}
+		}
+
+		log.Tracer(ctx).Debugf("changed Entry: %+v", entry)
+	}
+}
+
+//updateScoreIgnoreBlockPrevOffender updates this 4 Values of the Struct
+//ipString needs to correspond to the key of the entry in the map ips
+//WARNING: This function maybe deletes the entry ipString from the Map ips. (look at the returncode)
+//return: still in map? (bool)
+func (d *ipData) updateScoreIgnoreBlockPrevOffender(ipString string) bool {
+	d.score -= intMin(int(time.Since(d.lastUpdated)/decreaseInterval), d.score)
+
+	if d.ignore {
+		if time.Since(d.lastSeen) > unignoreTime {
+			d.ignore = false
+		}
+	}
+
+	if d.previousOffender && time.Since(d.lastSeen) > undoSuspicionIdleTime {
+		d.previousOffender = false
+	}
+
+	if d.blocked && time.Since(d.lastSeen) > unblockIdleTime {
+		d.blocked = false
+		d.blockedPorts = []tcpUDPport{}
+
+		status.DeleteThreat(threadPrefix + ipString)
+	}
+
+	if !d.blocked && !d.ignore && !d.previousOffender && d.score == 0 {
+		delete(ips, ipString)
+		return false
+	}
+
+	d.lastUpdated = time.Now()
+	return true
+}
+
+// Destroy implements the destroy interface.
+func (d *Detector) Destroy() error {
+	return nil
+}
+
+// DetectorFactory is a primitive detection method that runs within the factory only.
+func DetectorFactory(conn *network.Connection, pkt packet.Packet) (network.Inspector, error) {
+	return &Detector{}, nil
+}
+
+// Register registers the encryption detection inspector with the inspection framework.
+func init() {
+	ips = make(map[string]*ipData)
+
+	//cleanup old Threads
+	threads, _ := status.GetThreats(threadPrefix)
+	for _, t := range threads {
+		status.DeleteThreat(t.ID)
+	}
+
+	module = modules.Register("portscan-detection", nil, start, nil, "base", "netenv")
+	module.Enable()
```

Comment:
IMHO calling `status.GetThreads()` doesn't work at this point (or cannot expected to work). It requires the database module which is not yet brought up as we in the init and not in prepare/start phase.